synapse/models/dns.py

```diff
@@ -101,211 +107,76 @@ def getModelDefs(self):
-                    ('time', ('time', ()), {
-                        'ro': True,
-                        'doc': 'The time thel ookup occured.',
-                    }),
-                    ('client', ('inet:client', {}), {
-                        'doc': 'The client that requested the lookup.',
-                    }),
-                    ('client:ipv4', ('inet:ipv4', {}), {
-                        'doc': 'The IPv4 of the client that requested the lookup.',
-                    }),
-                    ('client:ipv6', ('inet:ipv6', {}), {
-                        'doc': 'The IPv6 of the client that requested the lookup.',
-                    }),
-                    ('client:port', ('inet:ipv4', {}), {
-                        'doc': 'The port of the client that requested the lookup.',
-                    }),
-                    ('server', ('inet:server', {}), {
-                        'doc': 'The server that responded to the lookup.',
-                    }),
-                    ('server:ipv4', ('inet:ipv4', {}), {
-                        'doc': 'The IPv4 of the server that responded to the lookup.',
-                    }),
-                    ('server:ipv6', ('inet:ipv6', {}), {
-                        'doc': 'The IPv6 of the server that responded to the lookup.',
-                    }),
-                    ('server:port', ('inet:ipv4', {}), {
-                        'doc': 'The port of the server that responded to the lookup.',
-                    }),
+
+                ('inet:dns:request', {}, (
+
+                    ('time', ('time', {}), {}),
+                    ('query', ('inet:dns:query', {}), {}),
+
+                    ('server', ('inet:server', {}), {}),
+
                     ('rcode', ('int', {}), {
-                        'doc': 'The DNS server response code.',
-                    }),
+                        'doc': 'The DNS server response code.'}),
+
                     ('exe', ('file:bytes', {}), {
-                        'doc': 'The file containing the code that attempted the DNS lookup.',
-                    }),
+                        'doc': 'The file containing the code that attempted the DNS lookup.'}),
+
                     ('proc', ('it:exec:proc', {}), {
-                        'doc': 'The process that attempted the DNS lookup.',
-                    }),
+                        'doc': 'The process that attempted the DNS lookup.'}),
+
                     ('host', ('it:host', {}), {
-                        'doc': 'The host that attempted the DNS lookup.',
-                    }),
-                    # one of the following should be set...
-                    ('a', ('inet:dns:a', {}), {
-                        'ro': True,
+                        'doc': 'The host that attempted the DNS lookup.'}),
+
+                )),
+
+                ('inet:dns:reply', {}, (
+
+                    ('ttl', ('int', {}), {}),
+                    ('request', ('inet:dns:request', {}), {}),
```

Comment:
Depending on the data source, you may not have a corresponding request for the reply (what if you only have visibility into one half of the conversation?). Are you assuming we can "reconstruct" the request from the reply? This is important because the only place in the inet:dns:reply to capture the IP of the responding DNS server is as part of the inet:dns:request secondary property.

Reply:
Agreed, in several data sources, we do only see the replies.  Fortunately, a reply is enough to reconstruct the request.  Thus the inet:dns:request node which always contains request data, and sometimes request & reply data ( rcode ).

Reply:
reconstructing the request assumes the dns server was well behaved.  that sounds like a risky thing to assume.

Reply:
Maybe. The only other option would be to split request into request/reply ( and rename reply to answer or something ) and hope to be able to tie them together sometimes? The trouble is, if we only created request ( and inet:dns:query ) for ones we observed directly ( and not the ones implied by server responses ) we would miss out on a *ton* of fused knowledge about client requests.